File: Core/src/com/stusherwin/midiapp/core/midi/modules/MidiToNotes2.java
package com.stusherwin.midiapp.core.midi.modules;

import com.stusherwin.midiapp.core.*;
import com.stusherwin.midiapp.core.midi.*;
import com.stusherwin.midiapp.core.modules.ThruModule;
import rx.Observable;
import rx.functions.Func1;
import rx.observables.GroupedObservable;

public class MidiToNotes2 extends ThruModule<MidiEvent, NoteWithDuration> {
    @Override
    public void init() throws InitializationException {
    }

    @Override
    public void destroy() {
    }

    @Override
    protected Observable<NoteWithDuration> transform(Observable<MidiEvent> input) {
        return input
                .filter(new Func1<MidiEvent, Boolean>() {
                    @Override
                    public Boolean call(MidiEvent midiEvent) {
                        return midiEvent instanceof NoteMidiEvent;
                    }
                })
                .cast(NoteMidiEvent.class)
                .groupByUntil(
                        new Func1<NoteMidiEvent, Integer>() {
                            @Override
                            public Integer call(NoteMidiEvent midiEvent) {
                                return midiEvent.getNote();
                            }
                        },
                        new Func1<GroupedObservable<Integer, NoteMidiEvent>, Observable<NoteMidiEvent>>() {
                            @Override
                            public Observable<NoteMidiEvent> call(GroupedObservable<Integer, NoteMidiEvent> integerNoteGroupedObservable) {
                                return integerNoteGroupedObservable.filter(new Func1<NoteMidiEvent, Boolean>() {
                                    @Override
                                    public Boolean call(NoteMidiEvent noteMidiEvent) {
                                        return noteMidiEvent instanceof NoteOff;
                                    }
                                });
                            }
                        }
                )
                .map(new Func1<GroupedObservable<Integer, NoteMidiEvent>, Observable<NoteOn>>() {
                    @Override
                    public Observable<NoteOn> call(GroupedObservable<Integer, NoteMidiEvent> integerNoteGroupedObservable) {
                        return integerNoteGroupedObservable
                                .filter(new Func1<NoteMidiEvent, Boolean>() {
                                    @Override
                                    public Boolean call(NoteMidiEvent noteMidiEvent) {

                                        return noteMidiEvent instanceof NoteOn;
                                    }
                                })
                                .cast(NoteOn.class);
                    }
                })
                .lift(new ToNoteWithDuration());
    }
}

File: Core/src/com/stusherwin/midiapp/core/midi/modules/MidiToNotes2.scala
package com.stusherwin.midiapp.core.midi.modules

import com.stusherwin.midiapp.core.midi._
import com.stusherwin.midiapp.core.modules.ThruModule
import rx.lang.scala.{Subject, Observable}
import com.stusherwin.midiapp.core.Note

class MidiToNotes2 extends ThruModule[MidiEvent, NoteWithDuration] {
  override def init() : Unit = {}
  override def destroy() : Unit = {}

  override def transform(input : Observable[MidiEvent]) : Observable[NoteWithDuration] = {
    val s = Subject.apply[NoteWithDuration]()

    input.filter(_.isInstanceOf[NoteMidiEvent])
         .map(n => n.asInstanceOf[NoteMidiEvent])
         .groupByUntil(
           evt => evt.getNote,
           (_ : Int, grp) => grp.filter(_.isInstanceOf[NoteOff]))
         .map({
           case (k, grp) => grp.filter(_.isInstanceOf[NoteOn])
                               .map(_.asInstanceOf[NoteOn]) })
         .subscribe(
           obs => {
             obs.subscribe(
               noteOn =>
                 s.onNext(new NoteWithDuration(
                   new Note(noteOn.getNote, noteOn.getVelocity),
                   obs.map(n => n.asInstanceOf[Object]))),
               (e : Throwable) => {},
               () => {})
           },
           (e:Throwable) => {},
           () => {})

    s
  }
}